tests/unit/optimization/StanFirstOrderSAGCostFunction_unittest.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Elfarouk
+ */
+
+#include <shogun/optimization/StanFirstOrderSAGCostFunction.h>
+#include "StanFirstOrderSAGCostFunction_unittest.h"
+#include <Eigen/Dense>
+#include <gtest/gtest.h>
+
+using namespace shogun;
+using Eigen::Matrix;
+using Eigen::Dynamic;
+using stan::math::var;
+using std::function;
+
+SGMatrix<float64_t> generateXSmall()
+{
+	SGMatrix<float64_t> X(2, 3);
+
+	X(0, 0) = 1;
+	X(0, 1) = 1;
+	X(0, 2) = 1;
+	X(1, 0) = 0;
+	X(1, 1) = 1;
+	X(1, 2) = 2;
+
+	return X;
+}
+
+StanVector generateParametersSmall()
+{
+	StanVector w(2, 1);
+	var m(1), c(0);
+	w(0, 0) = c;
+	w(1, 0) = m;
+
+	return w;
+}
+
+function<var(StanVector*, float64_t)>
+cost_for_ith_datapoint(SGMatrix<float64_t>& X, SGMatrix<float64_t>& y)
+{
+	auto f_i = [X, y](StanVector* w, int32_t idx) {
+		var wx_y = (*w)(0, 0) * X(0, idx) + (*w)(1, 0) * X(1, idx) - y(0, idx);
+		var res = wx_y * wx_y;
+		res /= 2;
```

Comment:
nitpicking, but this is the same as `res *= 0.5` except that the former is slower